symengine/lib/symengine_wrapper.pyx

```diff
@@ -1798,26 +1798,34 @@ cdef class DenseMatrixBase(MatrixBase):
         cdef DenseMatrixBase o = _sympify(rhs)
         if self.rows != o.rows:
             raise ShapeError("`self` and `rhs` must have the same number of rows.")
-        cdef DenseMatrixBase result = zeros(self.rows, self.cols + o.cols)
+        cdef DenseMatrixBase result = self.__class__(self)
+        deref(symengine.static_cast_DenseMatrix(result.thisptr)).resize(self.rows, self.cols + o.cols)
+        cdef Basic e_
         for i in range(self.rows):
             for j in range(self.cols):
-                result[i, j] = self[i, j]
+                e_ = _sympify(self._get(i, j))
+                deref(result.thisptr).set(i, j, e_.thisptr)
         for i in range(o.rows):
             for j in range(o.cols):
-                result[i, j + self.cols] = o[i, j]
+                e_ = _sympify(o._get(i, j))
+                deref(result.thisptr).set(i, j + self.cols, e_.thisptr)
         return result
 
     def col_join(self, bott):
         cdef DenseMatrixBase o = _sympify(bott)
         if self.cols != o.cols:
             raise ShapeError("`self` and `rhs` must have the same number of columns.")
-        cdef DenseMatrixBase result = zeros(self.rows + o.rows, self.cols)
+        cdef DenseMatrixBase result = self.__class__(self)
+        deref(symengine.static_cast_DenseMatrix(result.thisptr)).resize(self.rows + o.rows, self.cols)
```

Comment:
Change the above 2 lines to `cdef DenseMatrixBase result = self.__class__(self.rows, self.cols)`. This avoids the copy in the first line.